### PR TITLE
ddl, table: allow using SHARD_ROW_ID_BITS with auto_incremental columns

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
This PR allows using `SHARD_ROW_ID_BITS` for tables with `AUTO_INCREMENT` columns.

Before this PR, `SHARD_ROW_ID_BITS` cannot work with `AUTO_INCREMENT`:

```
tidb> CREATE TABLE T5 (`A` INT UNIQUE KEY AUTO_INCREMENT);
Query OK, 0 rows affected (0.01 sec)
tidb> ALTER TABLE T5 SHARD_ROW_ID_BITS = 4;
ERROR 1105 (HY000): unsupported shard_row_id_bits for table with auto_increment column.
```

This restriction leads to inconveniences to many TiDB users, this PR gives a quick way to remove it by:

* For `row_id`(or `handle`), TiDB still tries to make sharding if `SHARD_ROW_ID_BITS` is defined.
* For `auto_increment_id`(or `autoid`), TiDB doesn't make sharding to keep it incremental always.

### What is changed and how it works?
A new `AllocHandle()` is added to `table.Table`, which is used for the allocation of `handle`. Now `AllocAutoID()` will not try to do sharding for the allocated ids.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -916,6 +916,11 @@ func GetColDefaultValue(ctx sessionctx.Context, col *table.Column, defaultVals [
 
 // AllocAutoID implements table.Table AllocAutoID interface.
 func (t *tableCommon) AllocAutoID(ctx sessionctx.Context) (int64, error) {
```
Comment (coocood) [inline table/tables/tables.go]: Maybe `AllocAutoIncrementValue` is more accurate.

Comment (bb7133) [reply table/tables/tables.go]: addressed, thanks! @coocood

**Comment by coocood** [comment]
@bb7133 
We should check that the table's PK is not handle.

**Comment by bb7133** [comment]
> @bb7133
> We should check that the table's PK is not handle.

I didn't get your point: If the PK is handle, `AllocHandle()` is not called... the check is there already

**Comment by coocood** [comment]
> > @bb7133
> > We should check that the table's PK is not handle.
> 
> I didn't get your point: If the PK is handle, `AllocHandle()` is not called... the check is there already

I mean check it in the create table phase.

**Comment by bb7133** [comment]
addressed, PTAL @coocood @crazycs520

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -1219,6 +1219,9 @@ func buildTableInfoWithCheck(ctx sessionctx.Context, d *ddl, s *ast.CreateTableS
 	if err = handleTableOptions(s.Options, tbInfo); err != nil {
 		return nil, errors.Trace(err)
 	}
+	if tbInfo.ShardRowIDBits > 0 && tbInfo.PKIsHandle {
```
Comment (coocood) [inline ddl/ddl_api.go]: Why not put it in `handleTableOptions`?

Comment (bb7133) [reply ddl/ddl_api.go]: addressed, thanks!

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**